source/Calamari/Kubernetes/Conventions/HelmUpgradeWithKOSConvention.cs

```diff
@@ -0,0 +1,554 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Calamari.Common.Commands;
+using Calamari.Common.Features.Packages;
+using Calamari.Common.Features.Processes;
+using Calamari.Common.Features.Scripts;
+using Calamari.Common.FeatureToggles;
+using Calamari.Common.Plumbing.FileSystem;
+using Calamari.Common.Plumbing.Logging;
+using Calamari.Common.Plumbing.Variables;
+using Calamari.Deployment.Conventions;
+using Calamari.Kubernetes.Helm;
+using Calamari.Kubernetes.Integration;
+using Calamari.Kubernetes.ResourceStatus;
+using Calamari.Kubernetes.ResourceStatus.Resources;
+using Calamari.Util;
+using Newtonsoft.Json;
+using Octopus.CoreUtilities.Extensions;
+using YamlDotNet.RepresentationModel;
+
+namespace Calamari.Kubernetes.Conventions
+{
+    public class HelmUpgradeWithKOSConvention : IInstallConvention
+    {
+        readonly ILog log;
+        readonly ICommandLineRunner commandLineRunner;
+        readonly ICalamariFileSystem fileSystem;
+        readonly HelmTemplateValueSourcesParser valueSourcesParser;
+        readonly IResourceStatusReportExecutor statusReporter;
+        readonly IManifestReporter manifestReporter;
+        readonly Kubectl kubectl;
+
+        public HelmUpgradeWithKOSConvention(ILog log,
+                                            ICommandLineRunner commandLineRunner,
+                                            ICalamariFileSystem fileSystem,
+                                            HelmTemplateValueSourcesParser valueSourcesParser,
+                                            IResourceStatusReportExecutor statusReporter,
+                                            IManifestReporter manifestReporter,
+                                            Kubectl kubectl)
+        {
+            this.log = log;
+            this.commandLineRunner = commandLineRunner;
+            this.fileSystem = fileSystem;
+            this.valueSourcesParser = valueSourcesParser;
+            this.statusReporter = statusReporter;
+            this.manifestReporter = manifestReporter;
+            this.kubectl = kubectl;
+        }
+
+        public void Install(RunningDeployment deployment)
+        {
+            var isKOSEnabled = deployment.Variables.GetFlag(SpecialVariables.ResourceStatusCheck);
+
+            var releaseName = GetReleaseName(deployment.Variables);
+
+            var helmCli = new HelmCli(log, commandLineRunner, deployment);
+
+            kubectl.SetKubectl();
+
+            var currentRevisionNumber = helmCli.GetCurrentRevision(releaseName);
```

Comment:
We retrieve the current revision number, because we use the incremented value to retrieve the manifests